lib/tbot/service/supervisor.go

```diff
@@ -0,0 +1,258 @@
+// registered, or if the supervisor is already running.
+//
+// TODO: support registering services dynamically at-runtime.
+func (s *Supervisor) Register(service InternalService) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.started {
+		return trace.Errorf("cannot register services once the supervisor has started")
+	}
+
+	if _, ok := s.services[service.getName()]; ok {
+		return trace.Errorf("service %q is already registered", service.getName())
+	}
+
+	if !service.registerToSupervisor() {
+		return trace.Errorf("cannot register a service to more than one supervisor")
+	}
+
+	s.services[service.getName()] = service
+	return nil
+}
+
+// Run the registered services until the given context is canceled or a service
+// exits with an irrecoverable error.
+func (s *Supervisor) Run(ctx context.Context) error {
+	group, groupCtx := errgroup.WithContext(ctx)
+
+	s.mu.Lock()
+
+	if s.started {
+		s.mu.Unlock()
+		return trace.Errorf("cannot run a supervisor more than once")
+	}
+	s.started = true
+
+	for _, svc := range s.services {
+		svc := svc
+		group.Go(func() error {
+			return trace.Wrap(
+				s.superviseService(groupCtx, svc),
+				"service: %s", svc.getName(),
+			)
+		})
+	}
+	s.mu.Unlock()
+
+	err := group.Wait()
+	if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+		return nil
+	}
+	return err
+}
+
+// OneShot runs any registered services that implement OneShotHandler.
+func (s *Supervisor) OneShot(ctx context.Context) error {
+	// Note: we don't use errgroup here because we don't want one service's
+	// error to immediately cancel other services, in case they can gracefully
+	// degrade instead.
+	var wg sync.WaitGroup
+
+	s.mu.Lock()
+
+	if s.started {
+		s.mu.Unlock()
+		return trace.Errorf("cannot run a supervisor more than once")
+	}
+	s.started = true
+
+	errCh := make(chan error, len(s.services))
+	for _, svc := range s.services {
+		wg.Add(1)
+
+		svc := svc
+		go func() {
+			err := s.oneShotService(ctx, svc)
+			errCh <- trace.Wrap(err, "service: %s", svc.getName())
+			wg.Done()
+		}()
+	}
+	s.mu.Unlock()
+
+	wg.Wait()
+	close(errCh)
+
+	return trace.NewAggregateFromChannel(errCh, ctx)
+}
+
+func (s *Supervisor) superviseService(ctx context.Context, svc InternalService) error {
+	defer svc.finalize()
+
+	logger := s.logger.With("service", svc.getName())
+
+	setStatus := func(status status.Status) {
+		if svc.setStatus(status) {
+			logger.DebugContext(ctx, "Service status changed", "new_status", status)
+		}
+	}
+
+	retry, err := retryutils.NewRetryV2(retryutils.RetryV2Config{
+		Driver: retryutils.NewExponentialDriver(1 * time.Second),
+		Jitter: retryutils.HalfJitter,
+		Max:    1 * time.Minute,
+		Clock:  s.clock,
+	})
+	if err != nil {
+		return trace.Wrap(err, "creating retrier")
+	}
+
+	for {
+		start := time.Now()
+		err := svc.runHandler(ctx, &Runtime{setStatusFn: setStatus})
+		duration := time.Since(start)
+
+		setStatus(status.Failed)
```

Comment:
Is it always a failure, even if the `err == nil`?

Reply:
Have tried to make this clearer by returning an explicit error. In long-running mode, handlers are expected to run "forever" (i.e. until the context is canceled) so any early return is unexpected.